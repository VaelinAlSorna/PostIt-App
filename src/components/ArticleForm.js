import React from 'react';
import moment from 'moment';

class ArticleForm extends React.Component {
    state = {
        title: '',
        text: '',
        createdAt: moment().format('D. M. Y')
    };
    onTitleChange = (e) => {
        const title = e.target.value;
        this.setState(() => ({
            title
        }))
    }
    onTextChange = (e) => {
        const text = e.target.value;
        this.setState(() => ({
            text
        }))
    }
    onSubmit = (e) => {
        e.preventDefault();
        this.props.onSubmit(this.state)
    }
    render () {
       return ( 
        <div> 
            <form onSubmit={this.onSubmit}> 
                <input 
                    onChange={this.onTitleChange} 
                    placeholder="Title" 
                    type="text" 
                    value={this.state.title}
                />
                <textarea 
                    onChange={this.onTextChange}
                    placeholder="Add text for your article"
                    value={this.state.text}
                >
                </textarea>
                <input type="submit" value="Submit" />
            </form>
        </div>
       );
    }
}

export default ArticleForm;