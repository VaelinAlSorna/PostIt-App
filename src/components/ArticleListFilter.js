import React from 'react';
import { connect } from 'react-redux';
import { setYourArticlesFilter, removeYourArticlesFilter, setTextFilter } from '../actions/filterActions';

export class ArticleListFilter extends React.Component {
    handleSeeAll = () => {
            this.props.removeYourArticlesFilter();
            this.props.setTextFilter('');
    }
    handleSeeYours = () => {
        this.props.setYourArticlesFilter();
    }
    disablingBtnSeeAll = () => {
        if(this.props.filter.yourArticles || !!this.props.filter.text) {
            return false;
        } else {
            return true;
        }
    }
    disablingBtnSeeYours = () => {
        if(this.props.user.name && !this.props.filter.yourArticles) {
            return false;
        } else {
            return true;
        }
    }
    onTextFilterChange = (e) => {
        this.props.setTextFilter(e.target.value);
    }
    onSubmit = (e) => {
        //This is here in order to prevent user on mobile from sunmiting form
        e.preventDefault();
    }
    
    render () { 
        return  ( 
            <div className="filter">
                <form onSubmit={this.onSubmit} className="filter__group filter-form"> 
                    <input
                        className="filter-form__input"
                        onChange={this.onTextFilterChange} 
                        type="text" 
                        placeholder="Search Article By Name"
                        value={this.props.filter.text}
                    />
                </form>
                <button className="filter__group btn btn--green" onClick={this.handleSeeAll} disabled={this.disablingBtnSeeAll()}>See All Articles</button>
                <button className="filter__group btn btn--green" onClick={this.handleSeeYours} disabled={this.disablingBtnSeeYours()}>See Your Articles</button>
            </div>                
        )
    };
}

const mapStateToProps = (state) => ({
    filter: state.filter,
    user: state.auth
});

const mapDispatchToProps = (dispatch) => ({
    setYourArticlesFilter: () => dispatch(setYourArticlesFilter()),
    removeYourArticlesFilter: () => dispatch(removeYourArticlesFilter()),
    setTextFilter: (text) => dispatch(setTextFilter(text))
});

export default connect(mapStateToProps, mapDispatchToProps)(ArticleListFilter);

