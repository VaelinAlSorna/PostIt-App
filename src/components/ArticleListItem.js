import React from 'react';
//import { IoIosHeart, IoIosChatboxes } from "react-icons/io";
import { Link } from 'react-router-dom';


const ArticleListItem = ({ _id, title, createdAt, creator, comments, likes}) => (
    <Link to={`view/${_id}`}>
        <h3>{title}</h3>
        <p>Posted by: {creator} at: {createdAt}</p>
        {/*<p> <IoIosHeart /> {likes.length}</p>*/}
        {/*<p><IoIosChatboxes /> { comments.length }</p>*/}
    </Link>
);

export default ArticleListItem;