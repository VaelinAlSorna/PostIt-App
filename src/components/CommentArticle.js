import React from 'react';
import uuid from 'uuid';
import moment from 'moment';

class CommentArticle extends React.Component {
    state = {
        comment: '',
        createdAt: moment().format('D. M. Y'),
        error: ''
    };
    onTextareaChange = (e) => {
        const comment = e.target.value;
        this.setState(() => ({
            comment
        }));
    };
    onSubmit = (e) => {
        e.preventDefault();
        if (this.state.comment.trim().length < 3) {
            this.setState(() => ({
                error: 'Yout comment must contain at least 3 characters.'
            }));
        } else {
            const id = uuid();
            this.props.onSubmit({...this.state, id});
            this.setState(() =>  ({
                comment: '',
                error: ''
            }));
        }
    }
    render () {
        return (
            <form onSubmit={this.onSubmit}>
                { this.state.error && <p>{this.state.error}</p>}
                <textarea 
                    onChange={this.onTextareaChange} 
                    placeholder="What are your thoughts ?"
                    value={this.state.comment}
                >
                </textarea>
                <input type="submit" value="Comment" />
            </form>
        );
    };
};

export default CommentArticle;