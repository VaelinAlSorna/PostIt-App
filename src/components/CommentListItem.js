import React from 'react';

class CommentListItem extends React.Component {
   
    onDelete = () => {
        this.props.onClickDeleteComment(this.props.id);
    }
    render () { 
        return  (
                <li>
                    <p>{this.props.createdAt}</p>
                    <p>{this.props.comment}</p>
                    <button onClick={this.onDelete}>Delete</button>
                </li>
            )
    }
};

export default CommentListItem;