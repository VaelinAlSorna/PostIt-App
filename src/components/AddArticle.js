import React from 'react';
import { connect } from 'react-redux'
import ArticleForm from './ArticleForm';
import { addArticle } from '../actions/articlesActions'


class AddArticle extends React.Component {
    onSubmit = (article) => {
        this.props.dispatch(addArticle(article))
        this.props.history.push('/');
    }
    render () {
        return ( 
        <div>
            <ArticleForm onSubmit={this.onSubmit} />
        </div>
        );
    }
}

export default connect(undefined)(AddArticle);

