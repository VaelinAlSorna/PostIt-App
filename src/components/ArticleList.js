import React from 'react';
import { connect } from 'react-redux';
import ArticleListItem from './ArticleListItem';
import ArticleListFilter from './ArticleListFilter';
import yourArticlesFilter from '../selectors/yourArticlesFilter';

export const ArticleList = (props) => (
    <div className="main-content responsive-scroll">
        <ArticleListFilter />
        <ol className="list">
            {
            !!props.networkError
                ? <p className="server-error"> We are sorry ,but server is not responding ,please try to reconnect later.</p>   
                : (props.articles.length === 0) 
                    ? <p className="no-articles"> No articles fulfill these search conditions. </p>
                    : props.articles.map((article) => <ArticleListItem key={article._id} {...article} />)
            }
        </ol>
    </div>
);
 

const mapStateToProps = (state) => ({
    articles: yourArticlesFilter(state.articles, state.auth.id, state.filter),
    networkError: state.networkError
});

export default connect(mapStateToProps)(ArticleList);