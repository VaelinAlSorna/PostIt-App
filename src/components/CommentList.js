import React from 'react';
import CommentListItem from './CommentListItem';

const CommentList = (props) => (
         <ol>
            {
                props.comments.map((comment) => <CommentListItem 
                                                userId={props.userId}
                                                onClickDeleteComment={props.onClickDeleteComment} 
                                                key={comment._id} 
                                                {...comment}/>
                                            )
            }
        </ol>
)

export default CommentList;