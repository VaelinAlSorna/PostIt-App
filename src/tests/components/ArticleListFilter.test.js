import React from 'react';
import { shallow } from 'enzyme';
import { ArticleListFilter } from '../../components/ArticleListFilter';

test('should render ArticleListFilter with both buttons disabled set to true if user is not logged in', () => {
    const filter = {
        yourArticles: false,
        text: ''
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} user={{}}/>);
    expect(wrapper).toMatchSnapshot();
});

test('only button See Your Articles should not be disabled if user is logged in', () => {
    const filter = {
        yourArticles: false,
        text: ''
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} user={{name: 'Mark'}}/>);
    expect(wrapper.find('button').at(0).props().disabled).toBe(true);
    expect(wrapper.find('button').at(1).props().disabled).toBe(false);
});

test('both buttons should not be disabled if user is logged in and there is some text in search field', () => {
    const filter = {
        yourArticles: false,
        text: 'some text'
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} user={{name: 'Mark'}}/>);
    expect(wrapper.find('button').at(0).props().disabled).toBe(false);
    expect(wrapper.find('button').at(1).props().disabled).toBe(false);
});

test('only button See ALl Articles should not be disabled if user is logged in and he is alredy viewing his articles', () => {
    const filter = {
        yourArticles: true,
        text: 'some text'
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} user={{name: 'Mark'}}/>);
    expect(wrapper.find('button').at(0).props().disabled).toBe(false);
    expect(wrapper.find('button').at(1).props().disabled).toBe(true);
});

test('should call setTextFilter when value in input field changes', () => {
    const setTextFilter = jest.fn();
    const filter = {
        yourArticles: true,
        text: ''
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} setTextFilter={setTextFilter} user={{name: 'Mark'}}/>);
    wrapper.find('input').simulate('change', { target: { value: 'test' }});
    expect(setTextFilter).toHaveBeenLastCalledWith('test');
});

test('should call setYourArticlesFilter when user clicks See Your Articles button', () => {
    const setYourArticlesFilter = jest.fn();
    const filter = {
        yourArticles: true,
        text: ''
    }
    const wrapper = shallow(<ArticleListFilter filter={filter} setYourArticlesFilter={setYourArticlesFilter} user={{name: 'Mark'}}/>);
    wrapper.find('button').at(1).simulate('click');
    expect(setYourArticlesFilter).toHaveBeenCalled();
});

test('should call removeYourArticlesFilter and setTextFilter with empty string  when user clicks See All Articles button', () => {
    const removeYourArticlesFilter = jest.fn();
    const setTextFilter = jest.fn();

    const filter = {
        yourArticles: true,
        text: ''
    }
    const wrapper = shallow(<ArticleListFilter 
                                filter={filter} 
                                setTextFilter={setTextFilter} 
                                removeYourArticlesFilter={removeYourArticlesFilter} 
                                user={{name: 'Mark'}}
                            />);
    wrapper.find('button').at(0).simulate('click');
    expect(removeYourArticlesFilter).toHaveBeenCalled();
    expect(setTextFilter).toHaveBeenLastCalledWith('');
});



