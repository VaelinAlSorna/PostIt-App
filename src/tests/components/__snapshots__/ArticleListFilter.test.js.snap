// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ArticleListFilter with both buttons disabled set to true if user is not logged in 1`] = `
<div>
  <form>
    <input
      onChange={[Function]}
      placeholder="Search Article by Name"
      type="text"
      value=""
    />
  </form>
  <button
    disabled={true}
    onClick={[Function]}
  >
    See All Articles
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    See Your Articles
  </button>
</div>
`;
